---
title: "Format UCSC cellbrowser"
date: "`r Sys.Date()`"
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "html") })
output:
  html_document:
    toc: true
    toc_float: true
    toc_collapsed: false
    theme: cosmo
    highlight: tango
    fig_caption: true
    code_folding: hide
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


Format seurat objects into a UCSC cellbrowser. Note requires UCSC (`cellbrowser python package`[ https://cellbrowser.readthedocs.io/index.html]) 

```{r}
source(here::here("src", "utils.R"))
library(scbp)
library(tidyverse)
library(qs)
library(Seurat)
library(glue)
cb_outdir <- file.path("cellbrowser_v2")
dir.create(cb_outdir, showWarnings = FALSE, recursive = TRUE)
def_embeddings <- function(so) {str_subset(names(so@reductions), "umap")}

alra_objs <- file.path("objects", "alra")
dir.create(alra_objs, recursive = TRUE, showWarnings = FALSE)

write_config <- function(x, path){
  glue('{key}="{value}"', 
     key = names(x), 
     value = x) %>% 
    write_lines(path)
}

get_alra_assay <- function(so, file_name, overwrite = FALSE){

  ## only used to add to cellbrowser
  if(overwrite || !file.exists(file_name)){
    so <- SeuratWrappers::RunALRA(so, setDefaultAssay = FALSE)
    gc()
    alra_assay <- so@assays$alra
    qs::qsave(alra_assay, file_name)
  } else {
    alra_assay <- qs::qread(file_name)
  }
  alra_assay
}

```
# Collection overview

```{r}
collection_cb <- c(shortLabel="single cell RNA-seq profiling of immune populations in Low Grade Gliomas")

collection_desc <- c(title="single cell RNA-seq profiling of Low Grade Gliomas",
                     abstract="10x Genomics libraries were processed using Seurat. The cellbrowser contains one layouts, 'harmony_umap' which is a UMAP plot generated from running the harmony algorithm to correct for differences between tumors.  For the immune populations, ALRA imputated values can be obtained by querying for ALRA_genename, e.g. ALRA_HIF1A, will show the imputate values for HIF1A.",
                     unitDesc="Log-normalized counts")

write_config(collection_cb, file.path(cb_outdir, "cellbrowser.conf"))
write_config(collection_desc, file.path(cb_outdir, "desc.conf"))
```


## Myeloid cells


```{r}
so <- qread(file.path("objects", "so_myeloid.qs"))

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
   UPN = "UPN",
  `Dx` = "Dx",
  `Mutation Status` = "mutation_status",
  `Myeloid cell types` = "myeloid_cell_type",
  `Myeloid cell types with subclusters` = "myeloid_cell_type_subcluster",
  `Cell cycle phase` = "Phase")

alra <- get_alra_assay(so, file.path(alra_objs, "myeloid.qs"), overwrite = FALSE)
alra@data <- alra@data[rownames(so), ]
so@assays$alra <- alra

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                   "Phase",
                   "Dx",
                   "Mutation Status"), 
                 project = "myeloid",
                 outdir = cb_outdir,
                 marker_file = "markers/myeloid_cell_type_subcluster_cluster_markers.tsv",
                 ident = "Myeloid cell types with subclusters",
                 embeddings = c( "harmony_umap"),
                 skip_expr_matrix = TRUE,
                # assays = c("ALRA_" = "alra"),
                 config = list(priority = 1),
                 description = list(    
                   title = "Myeloid populations",
                   description = "10x genomics data"
                 )
)
```


## Myeloid cells


```{r}
so <- qread(file.path("objects", "so_lym.qs"))

cols_to_keep <- c(
  `genes per cell` = "nFeature_RNA",
  `UMIs per cell` = "nCount_RNA",
  `percent of UMIs as mitochondria` = "percent.mt",
   UPN = "UPN",
  `Dx` = "Dx",
  `Mutation Status` = "mutation_status",
  `T cell types` = "t_cell_type",
  `T cell types with subclusters` = "t_cell_type_subcluster",
  `Cell cycle phase` = "Phase")

alra <- get_alra_assay(so, file.path(alra_objs, "t.qs"), overwrite = FALSE)
alra@data <- alra@data[rownames(so), ]
so@assays$alra <- alra

make_cellbrowser(so, 
                 column_list = cols_to_keep,
                 secondary_cols = c(
                   "Phase",
                   "Dx",
                   "Mutation Status"), 
                 project = "t-cell",
                 outdir = cb_outdir,
                 marker_file = "markers/t_cell_type_subcluster_cluster_markers.tsv",
                 ident = "T cell types with subclusters",
                 embeddings = c( "harmony_umap"),
                 skip_expr_matrix = TRUE,
                 assays = c("ALRA_" = "alra"),
                 config = list(priority = 1),
                 description = list(    
                   title = "T cell populations",
                   description = "10x genomics data"
                 )
)
```

# Build all 

```{r}
cb_conf <- read_lines("~/.cellbrowser.conf") %>% 
  str_subset("dataRoot") %>% 
  str_split("=") %>% 
  .[[1]] %>% 
  .[2] %>% 
  gsub('\"', "", .)

if(cb_outdir != cb_conf ) {
  write_lines(paste0('dataRoot="', 
                     normalizePath(cb_outdir), '"'),
              "~/.cellbrowser.conf")
}
```

```{r}
datasets <- c(
   "myeloid",
   "t-cell"
)

datasets <- file.path(cb_outdir, datasets, "cellbrowser.conf")

build_cellbrowser(datasets, 
                  outdir = file.path(cb_outdir, "lgg-cellbrowser"),
                  cbBuild_path =  "/miniconda3/bin/cbBuild")
```



